Sparse softmax support (CUDA)

aocsa opened the pull request:
This PR implements softmax support for sparse tensors. 

Resolves gh-23651 for CUDA.

- [x]  sparse softmax
    - [x]  CUDA C++ implementation
    - [x]  unittests
    - [x]  update softmax documentation
    - [x]  autograd support
- [x]  sparse log_softmax
    - [x]  CUDA C++ implementation
    - [x]  unittests
    - [x]  update log_softmax documentation
    - [x]  autograd support
 
Here are some benchmark (script is [here](https://gist.github.com/aocsa/fbc1827b3e49901512a33ba96092cbc1)) results for `torch.sparse.softmax and torch.softmax`,  using CPU and GPU, values are float64 scalars, timing repeat is 1000:

| size         | density | sparse CUDA | sparse CPU |
|--------------|---------|-------------|------------|
|  (32, 10000) |   0.01  |    380.2    |    687.5   |
| (32, 10000)  | 0.05    | 404.3       | 2357.9     |
| (32, 10000)  | 0.1     | 405.9       | 3677.2     |
| (512, 10000) | 0.01    | 438.0       | 5443.4     |
| (512, 10000) | 0.05    | 888.1       | 24485.0    |
| (512, 10000) | 0.1     | 1921.3      | 45340.5    |


| size         | density | dense CUDA | dense CPU |
|--------------|---------|-------------|------------|
|  (32, 10000) |   0.01  |     23.6    |   1943.2   |
| (32, 10000)  | 0.05    | 23.6        | 1954.0     |
| (32, 10000)  | 0.1     | 23.5        | 1950.0     |
| (512, 10000) | 0.01    | 639.3       | 39797.9    |
| (512, 10000) | 0.05    | 640.3       | 39374.4    |
| (512, 10000) | 0.1     | 639.6       | 39192.3    |

Times are in microseconds (us).

Quick note:  I updated the performance test again.

rgommers commented on test/test_sparse.py:
```diff
@@ -2603,6 +2603,16 @@ def sparse_log(x):
                 ly1 = torch.sparse.log_softmax(x, dim)
                 self.assertEqual(ly1, sparse_log(y1))
 
+                if torch.cuda.is_available():
```
The outer `test_softmax` here is still decorated with `@cpu_only`, looks like that should be removed?

mruberry replied:
In the future it'd be nice if test_sparse.py added a device generic test class. Then tests could run on both CPU and CUDA.

mruberry commented:
Hey @aocsa, please ping me on this when it's ready for review.

aocsa commented:
> Hey @aocsa, please ping me on this when it's ready for review.
@mruberry  I updated this PR with the autograd support, I think it is ready to review.   cc @pearu

mruberry commented on test/test_sparse.py:
```diff
@@ -2386,7 +2386,7 @@ def test_sparse_to_numpy(self):
         self.assertRaises(TypeError, lambda: t.numpy())
 
-    @cpu_only
+    @skipIfRocm
```
In the future it'd be interesting to consider adding tests that validated the cpu and CUDA produced consistent results, too.

aocsa replied:
I think this test covers both cases as method `[_gen_sparse](https://github.com/pytorch/pytorch/blob/1ccdc91175e9900d9103081bb9d0f03c178ed694/test/test_sparse.py#L2599)`  uses `self.device` variable to generate sparse tensor with cpu and cuda configurations.

pearu commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,776 @@
+      "grad and output sparse dimensions must be equal");
+
+  AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "softmax_backward", [&] {
+    cuda_sparse_coo_softmax_backward<scalar_t, false>(
+        grad_input, grad, output, dim);
+  });
+  return grad_input;
+}
+
+Tensor log_softmax_backward_sparse_cuda(
+    const Tensor& grad_,
+    const Tensor& output_,
+    int64_t dim_,
+    const Tensor& input_) {
+  TensorArg grad_arg{grad_, "grad", 1}, output_arg{output_, "output", 2};
+  checkSameSize("softmax_backward", grad_arg, output_arg);
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+
+  auto grad = grad_.coalesce();
+  auto output = output_.coalesce();
+
+  Tensor grad_input = at::native::empty_like(output);
+  if (output.numel() == 0) {
+    return grad_input;
+  }
+
+  TORCH_CHECK(
+      dim >= 0 && dim < grad.dim(),
+      "dim must be non-negative and less than input dimensions");
+  TORCH_CHECK(
+      grad.sparse_dim() == output.sparse_dim(),
+      "grad and output sparse dimensions must be equal");
+
+  AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "softmax_backward", [&] {
+    cuda_sparse_coo_softmax_backward<scalar_t, true>(
+        grad_input, grad, output, dim);
+  });
+  return grad_input;
+}
```
This is repeating the functions in aten/src/ATen/native/sparse/SoftMax.cpp .

To avoid the code replication, I suggest updating the corresponding functions in aten/src/ATen/native/sparse/SoftMax.cpp with the following (example is given only for `softmax_sparse`, updates to other functions are analogous):
```
Tensor softmax_sparse(const Tensor& input_, const int64_t dim_, const bool half_to_float) {
  TORCH_INTERNAL_ASSERT(input_.is_sparse());
  TORCH_CHECK(!half_to_float, "softmax with half to float conversion is not supported on CPU");
  auto input = input_.coalesce();
  Tensor output = at::native::empty_like(input);
  if (input.numel() == 0) {
    return output;
  }
  TORCH_CHECK(dim_ >= 0 && dim_ < input.dim(),
              "dim must be non-negative and less than input dimensions");
  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "softmax", [&] {
      if (input.is_cuda()) {
         cuda_sparse_coo_softmax<scalar_t, false>(output, input, dim_);
      } else {
         cpu_sparse_coo_softmax<scalar_t, false>(output, input, dim_);
      }
    });
  return output;
}
```
Here `softmax_sparse` is the original `softmax_sparse_cpu` and the function `softmax_sparse_cuda` in the given file can be deleted.

After this renaming, update the native_functions.yaml accordingly:
```
- func: _sparse_softmax(Tensor self, int dim, bool half_to_float) -> Tensor
  use_c10_dispatcher: full
  dispatch:
    SparseCPU, SparseCUDA: softmax_sparse
```

aocsa replied:
I am going to use the device dispatcher to refactor this part of the code due to the cuda kernel code should live inside a .cu file.

pearu commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,776 @@
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<scalar_t*>>
+compute_pool_max(
+    const Tensor& indices,
+    const Tensor& values,
+    const IntArrayRef& sizes,
+    int64_t nvalues,
+    const int64_t dim) {
```
Please merge the template functions `get_pool` and `compute_pool_max` into a single template function that has an additional parameter, say `require_mx_rows`, about requesting the computation of `mx_rows`. This will reduce code replication in these functions considerably.

pearu reviewed (changes requested):
Thanks, @aocsa for implementing the CUDA support for softmax functions!

In my review, I noticed that the code size can be reduced considerably by eliminating code replication in a number of places. For instance, `get_pools` and `compute_pool_max` can be merged. Also, functions `softmax_sparse_cuda`, `log_softmax_sparse_cuda`, `softmax_backward_sparse_cuda`, and `log_softmax_backward_sparse_cuda` can be deleted after updating the corresponding functions in aten/src/ATen/native/sparse/SoftMax.cpp .

In addition, I am curious about the implementation complexity of `compute_pool_max` for computing `mx_rows` when for CPU it requires only 6 lines of code.

pearu commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,776 @@
+get_pools(
+    const Tensor& indices,
+    const IntArrayRef& sizes,
+    int64_t nvalues,
```
nevermind when merging with compute_pool_max

aocsa commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,776 @@
+      sorted_indices, device_pool_offsets, count_set, mx_rows);
+}
+
+template <typename scalar_t, bool is_log_softmax>
```
all right

pearu replied:
I believe the consistent naming of variables simplifies reading and maintaining the code. Why did you ignore it?

pearu replied:
Nevermind! It was my fault not to properly update my emacs view.

aocsa commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,776 @@
+        compute_pool_max<scalar_t>(indices, values_cp, sizes, nvalues, dim);
+
+    auto pool_size = device_pool_offsets.size();
+    thrust::device_vector<scalar_t> exp_sums_rows(nvalues * pool_size, 0);
+    int block_size = getNumThreads(pool_size);
+    const int grid_size = (pool_size + block_size - 1) / block_size;
+
+    cuda_sparse_coo_softmax_kernel<scalar_t, is_log_softmax>
+        <<<grid_size, block_size, 0, stream>>>(
+            sorted_indices.data().get(),
+            pool_size,
+            device_pool_sizes.data().get(),
+            device_pool_offsets.data().get(),
+            nvalues,
+            mx_rows.data().get(),
+            exp_sums_rows.data().get(),
+            values_accessor,
+            out_values_accessor);
+  } else {
+    std::tie(sorted_indices, device_pool_offsets, device_pool_sizes, mx_rows) =
+        compute_pool_max<scalar_t>(indices, values_2, sizes, nvalues, dim);
+
+    auto pool_size = device_pool_offsets.size();
+    thrust::device_vector<scalar_t> exp_sums_rows(nvalues * pool_size, 0);
+    int block_size = getNumThreads(pool_size);
+    const int grid_size = (pool_size + block_size - 1) / block_size;
+
+    cuda_sparse_coo_softmax_kernel<scalar_t, is_log_softmax>
+        <<<grid_size, block_size, 0, stream>>>(
+            sorted_indices.data().get(),
+            pool_size,
+            device_pool_sizes.data().get(),
+            device_pool_offsets.data().get(),
+            nvalues,
+            mx_rows.data().get(),
+            exp_sums_rows.data().get(),
+            values_accessor,
+            out_values_accessor);
+  }
+}
```
Yes, the mx_rows is just an index reference to the values tensor and the mx_rows values are only updated when LogSoftMax is enabled, so a new copy is needed. 

Thanks for the suggestion, I tried to do that before, however, it looks like allocating many small buffers inside a cuda kernel is slower than allocating all needed at the beginning. Maybe using shared_memory inside the kernel but the algorithm would be different.

aocsa replied:
All right!

aocsa commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,776 @@
+      auto out_values_row = ouput_values_acc[i];
+      for (int64_t j = 0; j < nvalues; j++) {
+        if (LogSoftMax) {
+          out_values_row[j] = values_row[j] - mx_row[j];
```
All right, let me try this approach.

aocsa replied:
All right!

aocsa commented:
> Thanks, @aocsa for implementing the CUDA support for softmax functions!
> 
> In my review, I noticed that the code size can be reduced considerably by eliminating code replication in a number of places. For instance, `get_pools` and `compute_pool_max` can be merged. Also, functions `softmax_sparse_cuda`, `log_softmax_sparse_cuda`, `softmax_backward_sparse_cuda`, and `log_softmax_backward_sparse_cuda` can be deleted after updating the corresponding functions in aten/src/ATen/native/sparse/SoftMax.cpp .
> 
Thanks for the review @pearu, I reduced the code, refactoring it where it was required.  I tried to use the `device dispatcher` to reduce common code between cpu and cuda code, however it seems that it only works with specific paths. Look at  this [TODO](https://github.com/pytorch/pytorch/blob/9d1138afec26a4fe0be74187e4064076f8d45de7/aten/src/ATen/native/DispatchStub.h#L35). So I just did a simple refactor using what you suggested.

> In addition, I am curious about the implementation complexity of `compute_pool_max` for computing `mx_rows` when for CPU it requires only 6 lines of code.

The complexity of compute_pool_max function should be in fact faster as it computes the `pools` and `maxes` in parallel using `thrust::reduce_by_key` API. However, it pays with the cost of the pre-processing step, using `thrust::sorting`, and its corresponding temporal memory uses.    

cc @mruberry

pearu commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,776 @@
+  }
+  return MAX_BLOCK_SIZE;
+}
+
```
Can you comment on why the current approach is preferred?

pearu replied:
Nevermind! I was viewing outdated emacs buffer..

pearu commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,776 @@
+    int64_t offset = pool_offsets[index];
+    int64_t* pool_indices = sorted_pool_indices + offset;
+    int64_t pool_indices_size = pool_sizes[index];
+    /* Compute mx */
```
Unattended (again). Please delete the comment, it does not make sense.

pearu replied:
Nevermind!

pearu commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,645 @@
+      log_softmax_max_buffer.resize(new_sz * nvalues);
+      auto log_softmax_max_buffer_ptr = log_softmax_max_buffer.data().get();
+      for (int64_t index = 0; index < new_sz; index++) {
+        auto current = mx_rows_indices[index];
+        cudaMemcpy(
+            log_softmax_max_buffer_ptr + index * nvalues,
+            values.data_ptr<scalar_t>() + index * nvalues,
+            nvalues * sizeof(scalar_t),
+            cudaMemcpyDeviceToDevice);
+      }
```
I think you have a bug here:
- `current` is not used
- `log_softmax_max_buffer` will not contain the max values

In theory, the max values are not necessary for computing softmax, these only provide numerically more stable results. Hence, the softmax tests pass even when the used max values are wrong. As a result, the softmax on CUDA will be less stable than the softmax on CPU when using extreme inputs.

aocsa replied:
Thanks for notice me about this issue. I fixed in a new commit. Basically what I am doing is to create a buffer that contains just the  max values `log_softmax_max_buffer` that is going to be needed and updated in the cuda-kernel when `LogSoftMax` is enabled. 

```
cudaMemcpy(
            log_softmax_max_buffer_ptr + index * nvalues,
            values.data_ptr<scalar_t>() + current * nvalues,
            nvalues * sizeof(scalar_t),
            cudaMemcpyDeviceToDevice);
```
So in this line, I am copying these values. Then, here, I am creating the references at `mx_rows` using this new buffer. 
```
thrust::transform(
          thrust::make_counting_iterator(int64_t(0)),
          thrust::make_counting_iterator(int64_t(new_sz)),
          mx_rows.begin(),
          [log_softmax_max_buffer_ptr, nvalues] __device__(int64_t index) {
            return (scalar_t*)log_softmax_max_buffer_ptr + index * nvalues;
          }); 
```

pearu commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,645 @@
+          device_pool_sizes.data().get(),
+          device_pool_offsets.data().get(),
+          nvalues,
+          mx_rows.data().get(),
```
Shouldn't `log_softmax_max_buffer` be used here instead of `mx_rows`?

aocsa replied:
As `mx_rows` is a valid index reference to `log_softmax_max_buffer`, no changes is required here.

pearu reviewed (changes requested):
@aocsa thanks for addressing my comments! Great work!

I believe there is a bug regarding the computation of `log_softmax_max_buffer` and this does not appear to be used, IIRC.

pearu commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,645 @@
+      auto log_softmax_max_buffer_ptr = log_softmax_max_buffer.data().get();
+      for (int64_t index = 0; index < new_sz; index++) {
+        auto current = mx_rows_indices[index];
+        cudaMemcpy(
+            log_softmax_max_buffer_ptr + index * nvalues,
+            values.data_ptr<scalar_t>() + current * nvalues,
+            nvalues * sizeof(scalar_t),
+            cudaMemcpyDeviceToDevice);
```
This is incorrect for `nvalues>1`.


(i) The corresponding code has to be equivalent to https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/SoftMax.cpp#L330-L335

(ii) Consider the following example:
```
indices = [[0, 0, 0, 1, 1], [0, 3, 4, 1, 3]]
s = 100
values = [[11*s, -11*s], [14*s, -14*s], [15*s, -15*s], [22*s, -22*s], [24*s, -24*s]]

for device in ['cpu', 'cuda']:
    X = torch.sparse_coo_tensor(indices, values, (2, 5, 2), dtype=float, device=device)
    Y = torch.sparse.log_softmax(X, dim=0)
    print(Y)
```
that outputs
```
tensor(indices=tensor([[0, 0, 0, 1, 1],
                       [0, 3, 4, 1, 3]]),
       values=tensor([[    0.,     0.],
                      [-1000.,     0.],
                      [    0.,     0.],
                      [    0.,     0.],
                      [    0., -1000.]]),
       size=(2, 5, 2), nnz=5, dtype=torch.float64, layout=torch.sparse_coo)
tensor(indices=tensor([[0, 0, 0, 1, 1],
                       [0, 3, 4, 1, 3]]),
       values=tensor([[    0.,     0.],
                      [-1000.,   -inf],
                      [    0.,     0.],
                      [    0.,     0.],
                      [    0.,   -inf]]),
       device='cuda:0', size=(2, 5, 2), nnz=5, dtype=torch.float64,
       layout=torch.sparse_coo)
```
where the result from cuda target is wrong. Try `s=10` to get matching results.

(iii) When dumping the content of `log_softmax_max_buffer`, it is
```
log_softmax_max_buffer=[1100 -1100 2200 -2200 2400 -2400 1500 -1500 ]
```
but the output should be
```
log_softmax_max_buffer=[1100 -1100 2200 -2200 2400 -1400 1500 -1500 ]
```

aocsa replied:
Thanks for notice me about this case. The example was really helpful to understand the error in the code. I updated the way the max values are computed.

pearu reviewed (changes requested):
@aocsa thanks, we are almost there... but the code computing the max values is unfortunately incorrect. See the given example and diagnostics below.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3363,12 +3363,12 @@
   use_c10_dispatcher: full
   dispatch:
-    SparseCPU: softmax_sparse_cpu
+    SparseCPU, SparseCUDA: softmax_sparse
```
I think you're going to watch to separate these dispatches (here and for the other functions)

aocsa replied:
all right!

mruberry commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -531,13 +630,15 @@ Tensor log_softmax_sparse_cpu(const Tensor& input_, const int64_t dim_, const bo
-  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "log_softmax", [&] {
-      cpu_sparse_coo_softmax<scalar_t, true>(output, input, dim_);
-  });
+  if (input.is_cuda()) {
```
I think you want to leave this dispatch logic to be autogenerated from the yaml entries. What, after all, if CUDA is not installed? log_softmax_sparse_cuda won't be available.

aocsa replied:
That is a good point, and that one of  the reason to use the device dispatcher if we want to avoid duplicate code. As I explained before, I tried to use in many ways the device dispatcher in the sparse module to reduce common code between cpu and cuda code, however it seems that it only works with specific paths. Look at this [TODO](https://github.com/pytorch/pytorch/blob/9d1138afec26a4fe0be74187e4064076f8d45de7/aten/src/ATen/native/DispatchStub.h#L35). 

So I think that in order to maintain compatibility even when CUDA is not enabled I will go back and leave the dispatch logic to be autogenerated from yaml entries. cc @pearu

mruberry commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -505,7 +506,102 @@ void cpu_sparse_coo_softmax_backward(Tensor& grad_input, const Tensor& grad, con
 } // namespace
 
-Tensor softmax_sparse_cpu(const Tensor& input_, const int64_t dim_, const bool half_to_float) {
+void softmax_sparse_cpu(
```
Maybe not in this PR, but should there eventually be native/sparse/cpu, analogous to aten/native/cpu vs aten/native/cuda?

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,645 @@
+  return threadSizes[4];
+}
+
+int64_t get_nvalues(const IntArrayRef& sizes, int64_t sparse_dim) {
```
Is there a more descriptive name for this function? It is, admittedly, tricky to describe. It's computing the number of elements in each dense (sub)tensor, right?

aocsa replied:
All right, Yes, it does. It is just  the number of entries in the dense part of a sparse tensor.

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,645 @@
+namespace {
+
+// Number of threads in a block given an input size up to MAX_BLOCK_SIZE
+static int getNumThreads(int nElem) {
```
Any reason why these particular values were chosen?

aocsa replied:
As usual, we are choosing  the closest power of 2 near to the problem size. The MAX_BLOCK_SIZE is because it matches with the maximum num of threads in dimension X in most configuration, compute capability >= 2.0.

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,645 @@
+}
+
+template <typename scalar_t, bool LogSoftMax>
+__global__ void cuda_sparse_coo_softmax_kernel(
```
Might as well add a comment describing each of these functions

pearu replied:
Or as a minimum, reference the corresponding functions in CPU target that document the algorithms in detail.

ngimel commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,645 @@
+      auto log_softmax_max_buffer_ptr = log_softmax_max_buffer.data().get();
+      for (int64_t index = 0; index < new_sz; index++) {
+        auto current = mx_rows_indices[index];
+        cudaMemcpy(
```
please don't use synchronizing cuda APIs. The best way would be to use pytorch's APIs to copy tensors, if this is undesirable for some reason, use cudaMemcpyAsync and specify stream it has to run on.

aocsa replied:
all right!

ngimel commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,645 @@
+    }
+    for (int64_t j = 0; j < nvalues; j++) {
+      if (LogSoftMax) {
+        mx_row[j] += c10::cuda::compat::log(exp_sums_row[j]);
```
This is numerically unstable, see https://github.com/pytorch/pytorch/commit/c34b333230a4d56cb7fc246a094a050b20e73aa6

pearu replied:
To be accurate, it is numerally less stable for data with extreme differences, otherwise one would not notice the effect.
But I agree stability can be improved by using here
```
exp_sums_row[j] = c10::cuda::compat::log(exp_sums_row[j]);
```
and below use
```
out_values_row[j] = values_row[j] - mx_row[j] - exp_sums_row[j];
```
This is even better because `mx_row` is not modified inplace and then log_softmax_max_buffer` can be eliminated.

aocsa replied:
ohhh nice to know it.

ngimel commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,597 @@
+      host_strides[i] = host_strides[i + 1] * (i + 1 == dim ? 1 : sizes[i + 1]);
+    }
+  }
+  thrust::device_vector<int64_t> strides = host_strides;
```
thrust::device_vector would use thrust's default allocator (cudaMalloc/cudaFree), those calls are synchronizing and expensive.

aocsa committed:
sparse softmax (CUDA). Closes gh-23651

minor fix

Initial commit: cuda_sparse_softmax and cuda_sparse_log_softmax


minor redo

minor fix

Closes gh-23651

update with huge performance improvement!


autograd support

minor fix to autograd support


minor refactor


update with minor refactor 


optimize LogSoftMax mx buffer


minor bug solved

fix max_values error


updates based on change requests

updates based on change requests


minor update

aocsa commented:
Thanks for the review @pearu, @ngimel, @mruberry. I added some documentation and references to the algorithms that this  implementation is based on. 

Besides that, main updates here are:
1)  Solve the issue with  max values, which was reported by @pearu 
2)  Numerically is more stable for backward algorithm. 
3)  Lastly, instead using thrust::device_vector now we are using at::Tensor class to represent intermediate data.

aocsa committed:
minor redo

aocsa commented:
I minor update,  I solved some problems  reported by CI, now all the checks are green. @ngimel, @mruberry this PR is ready to look it over. Thanks in advance, cc @pearu

pearu reviewed (changes requested):
Thanks, @aocsa, looks good!

I have some nit-comments: reduce code size and unifies the naming of variables.

aocsa committed:
minor update: reduce code size and unifies the naming of variables.

aocsa commented:
> reduce code size and unifies the naming of variables. 

Thanks @pearu, I updated this PR based on your feedback.  
cc  @ngimel, @mruberry

pearu reviewed (approved):
LGTM. Thanks, @aocsa !

mruberry commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -429,4 +429,4 @@ set(ATen_HIP_INCLUDE ${ATen_HIP_INCLUDE} PARENT_SCOPE)
 set(ATen_CPU_DEPENDENCY_LIBS ${ATen_CPU_DEPENDENCY_LIBS} PARENT_SCOPE)
 set(ATen_CUDA_DEPENDENCY_LIBS ${ATen_CUDA_DEPENDENCY_LIBS} PARENT_SCOPE)
-set(ATen_HIP_DEPENDENCY_LIBS ${ATen_HIP_DEPENDENCY_LIBS} PARENT_SCOPE)
+set(ATen_HIP_DEPENDENCY_LIBS ${ATen_HIP_DEPENDENCY_LIBS} PARENT_SCOPE)
```
newline - looks like this was accidentally modified?

aocsa replied:
Yes, It was an accident. I will recover it back

mruberry commented on aten/src/ATen/native/sparse/utils/ParamUtils.h:
```diff
@@ -0,0 +1,57 @@
+  TORCH_CHECK(
+      !half_to_float,
+      std::string(function_name) +
+          ": with half to float conversion is not supported on CPU");
```
this references the CPU but is checked by both CPU and CUDA, right? Maybe "... half to float conversion is not supported for sparse inputs"?

aocsa replied:
Yes you are right, so now it is:
`TORCH_CHECK(!half_to_float, std::string(function_name) + ": with half to float conversion is not supported on " + input_.device().str());`

About last question, half_to_float is not supported even with dense tensors, see here: https://github.com/pytorch/pytorch/blob/bcb0bb7e0e03b386ad837015faba6b4b16e3bfb9/aten/src/ATen/native/SoftMax.cpp#L121
I am now sure why.

mruberry commented on aten/src/ATen/native/sparse/utils/ParamUtils.h:
```diff
@@ -0,0 +1,57 @@
+      !half_to_float,
+      std::string(function_name) +
+          ": with half to float conversion is not supported on CPU");
+  auto input = input_.coalesce();
```
It's strange that the input is coalesced and the output constructed in a function called "check_invariants"

mruberry replied:
(seems like a function named "check_invariants" should return void or bool)

mruberry replied:
The easiest fix may be renaming this function (and the next one)

mruberry commented on aten/src/ATen/native/sparse/utils/ParamUtils.h:
```diff
@@ -0,0 +1,57 @@
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+
+  auto grad = grad_.coalesce();
```
Same comments here as above

mruberry commented on aten/src/ATen/native/sparse/utils/ParamUtils.h:
```diff
@@ -0,0 +1,57 @@
+} // namespace apply
+
+} // namespace native
+} // namespace at
```
newline

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+    PackedTensorAccessor<scalar_t, 2> input_values_acc,
+    PackedTensorAccessor<scalar_t, 2> ouput_values_acc) {
+  /*
+    See ATen/native/sparse/cpu/Softmax.cpp:cpu_sparse_coo_softmax for the CPU
```
ATen/native/sparse/Softmax.cpp

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+    const Tensor& input,
+    const int64_t dim) {
+  /*
+    See ATen/native/sparse/cpu/Softmax.cpp:cpu_sparse_coo_softmax for the CPU
```
sparse/Softmax.cpp

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+  auto values = input._values().contiguous();
+  auto out_values = output._values();
+  auto out_indices = output._indices();
+  out_values.resize_as_(values);
```
Is output already the correct size due to the "check_invariants" call?

aocsa replied:
not at this point, 'check_invariants' only check and allocates output tensor as `::empty_like(input)`.

aocsa replied:
So, it is needed as `empty_like` for sparse tensor just init the sparse layout like the input but not the values and indices.

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+  if (dim >= sparse_dim) {
+    if (LogSoftMax) {
+      auto new_values = log_softmax_cuda(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
```
Can you use `set_` to avoid a copy here?

aocsa replied:
Yes, nice to know about set_ method.

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+      out_values.copy_(new_values);
+    } else {
+      auto new_values = softmax_cuda(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
```
Same question re: `set_` here

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+  auto nvalues = get_nvalues(sizes, sparse_dim);
+
+  /* Prepare accessors */
+  auto values_2 = values.view({nnz, nvalues});
```
can nvalues be computed as values.numel()/nnz?

aocsa replied:
yes, you are right (y)

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+    Return pools of indices that align with the given dimension and the
+    correspoinding max values for each pool.
+
+    See ATen/native/sparse/cpu/Softmax.cpp:get_offsets and
```
sparse/Softmax.cpp

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+    const int64_t dim) {
+  /*
+    Return pools of indices that align with the given dimension and the
+    correspoinding max values for each pool.
```
corresponding

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+    const IntArrayRef& sizes,
+    const int64_t dim) {
+  /*
+    See ATen/native/sparse/cpu/Softmax.cpp:get_offsets for the CPU
```
sparse/Softmax.cpp

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+
+  auto ndim = indices.size(0);
+  auto nnz = indices.size(1);
+  auto host_strides =
```
Creating a CPU tensor here seems excessive. What about creating a vector and using something like std::accumulate in place of the loop below?

mruberry replied:
(you probably need to use `from_blob` to construct a tensor from a vector)

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+      [offsets_ptr] __device__(int64_t x, int64_t y) {
+        return offsets_ptr[x] == offsets_ptr[y];
+      });
+  auto new_sz = thrust::distance(
```
Can this just subtract the pointers to compute the difference instead of calling thrust::distance?

aocsa replied:
It could be but with thrust the most common way to do that is using thrust::distance especially if we work with have to deal with `thrust::device_ptr`.

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+      [offsets_ptr] __device__(int64_t x, int64_t y) {
+        return offsets_ptr[x] < offsets_ptr[y];
+      });
+  auto pool_sizes = at::empty({nnz}, indices.options());
```
Is there a more accurate pool size that can be computed here? That is, does this always need to upper bound on nnz?

aocsa replied:
Yes, we begin with nnz but then the pool size is resized once we know the number of pools inside:  `pool_sizes.resize_({new_sz});`

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+
+  Tensor mx_buffer;
+  if (requireMxRows) {
+    mx_buffer = at::empty({new_sz * nvalues}, values.options());
```
Why not just create the tensor filled with the correct value (of negative infinity)?

aocsa replied:
All right, so for the best of my knowledge this has to be done in two steps 
```
    auto output = at::empty(new_shape, self.options());
    output.fill_(value);
```
I will change using this approach. If there is a better way please let me know, cc @mruberry

mruberry replied:
Can you use `full`? https://pytorch.org/docs/master/generated/torch.full.html?highlight=full#torch.full

aocsa replied:
Yes I can

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+  Tensor mx_buffer;
+  if (requireMxRows) {
+    mx_buffer = at::empty({new_sz * nvalues}, values.options());
+    auto values_accessor =
```
It would be clearer to create this closer to where it's used

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+      compute_pool_max<scalar_t, true>(indices, values_2, sizes, nvalues, dim);
+
+  auto pool_size = pool_offsets.size(0);
+  auto exp_sums_rows = at::empty({nvalues * pool_size}, values.options());
```
Instead of filling can this tensor just be created with at::zeros? Or, even better, if these values are only written to and not read does this need to be filled at all?

aocsa replied:
Yes, at::zeros is better, and as this buffer data is used as an accumulator inside the kernel, the initialization with zero is needed.

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+    scalar_t* mx_row = mx_rows + index * nvalues;
+
+    /* Apply exp to (v - mx) and sum the results */
+    scalar_t* exp_sums_row = exp_sums_rows + index * nvalues;
```
Seems like exp sums can be accumulated in registers and then the final value written out; this might let you start with exp_sums_row = 0 instead of reading it (see previous comment about filling vs. leaving exp_sums_row uninitialized)

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+    }
+    for (int64_t j = 0; j < nvalues; j++) {
+      if (LogSoftMax) {
+        exp_sums_row[j] = c10::cuda::compat::log(exp_sums_row[j]);
```
following the above, this should minimize writes to global memory by writing to a register (local) value first and then writing only the final values to global memory

aocsa replied:
All right! I got it!

mruberry replied:
Sorry to be so repetitive, we were just trying to mark ideas as we went through for clarity.

aocsa replied:
Don't worry and thanks Mike, It nice to know  this  optimizations can be done. I've just tried to understand well until I got it the idea. 
So, I updated the PR covering most of the feedback, tomorrow I will update the benchmark and add new unittests. Thanks again :)

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+      auto out_values_row = ouput_values_acc[i];
+      for (int64_t j = 0; j < nvalues; j++) {
+        if (LogSoftMax) {
+          out_values_row[j] = values_row[j] - mx_row[j] - exp_sums_row[j];
```
out_values_row[j], in particular, seems to be getting hammered with writes

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,680 @@
+        if (LogSoftMax) {
+          out_values_row[j] = values_row[j] - mx_row[j] - exp_sums_row[j];
+        } else {
+          out_values_row[j] *= exp_sums_row[j];
```
getting rid of these global reads is likely to improve performance

mruberry replied:
(out_values_row[j] is also written to multiple times in this case)

mruberry commented:
Hey @aocsa, I had the opportunity to review forward with @ngimel and we made some suggestions and asked a few questions. We plan to go through backward soon.

We have a couple broader questions, too:

- For perf, could we see examples with 32x10,000 and 512x10,000 (all sparse) with varying densities: 1%, 5% and 10%, also comparing with SparseCPU and DenseCUDA implementations?
- For testing, would you add test validating the consistency of SparseSoftmax on CPU and CUDA by comparing the results to DenseSoftmax on CPU and CUDA? 

In general, as we move forward with Sparse, it might be worthwhile to invest in making (grad)gradcheck work for sparse ops, updating the tests to be more granular (and use the device generic framework), and automatically compare consistency with the corresponding dense ops.

aocsa committed:
Step1: Minor refactor and improvements

aocsa committed:
Step2: Minor refactor and improvements

aocsa commented:
This PR is updated accordingly to the feedback and request changes.

I updated the performance test, see script [here](https://gist.github.com/aocsa/fee35c465e6aa2f2316e88a1b1a77ab9). Moreover,  I did new tests for 32x10,000 and 512x10,000 sparse tensors with varying densities: 1%, 5% and 10% comparing with SparseCPU and DenseCUDA. Here the results: 

| sizes        | sparsity | dimension | pytorch dense(CPU) | pytorch dense (CUDA) | pytorch sparse (CPU) | pytorch sparse(CUDA) |
|--------------|----------|-----------|--------------------|----------------------|----------------------|----------------------|
| (32, 10000)  | 0.01     | 0         |  200.94 us         |  9.48 us             |  1101.36 us          |  443.06 us           |
| (32, 10000)  | 0.01     | 1         |  62.28 us          |  8.66 us             |  298.90 us           |  539.61 us           |
| (32, 10000)  | 0.05     | 0         |  203.93 us         |  8.56 us             |  3278.01 us          |  398.45 us           |
| (32, 10000)  | 0.05     | 1         |  51.34 us          |  8.15 us             |  1169.31 us          |  1076.84 us          |
| (32, 10000)  | 0.1      | 0         |  207.54 us         |  8.57 us             |  5921.73 us          |  418.67 us           |
| (32, 10000)  | 0.1      | 1         |  61.16 us          |  8.17 us             |  3120.59 us          |  1781.29 us          |
| (512, 10000) | 0.01     | 0         |  2986.89 us        |  8.31 us             |  8611.98 us          |  450.85 us           |
| (512, 10000) | 0.01     | 1         |  390.07 us         |  7.89 us             |  4535.45 us          |  1345.43 us          |
| (512, 10000) | 0.05     | 0         |  2882.00 us        |  8.24 us             |  29257.63 us         |  887.44 us           |
| (512, 10000) | 0.05     | 1         |  388.50 us         |  7.95 us             |  19558.99 us         |  3184.66 us          |
| (512, 10000) | 0.1      | 0         |  2139.92 us        |  8.66 us             |  51898.52 us         |  2008.65 us          |
| (512, 10000) | 0.1      | 1         |  513.24 us         |  7.83 us             |  37152.97 us         |  6238.08 us          |

Some notes:
- cuda softmax performance is better now with the new improvements like minimizing writes to global memory. Compare with previous results, now it consumes less memory and it is little faster from 20X to 21.5X

> For testing, would you add test validating the consistency of SparseSoftmax on CPU and CUDA by comparing the results to DenseSoftmax on CPU and CUDA?

About this, the test covers this case, see here `softmax_to_dense` https://github.com/pytorch/pytorch/blob/e931883e121dc3fd58e3b4b77e29afc310dbc8d1/test/test_sparse.py#L2610
It compares the consistency of the result on both CPU and CUDA with torch.sparse.softmax as `test_op>_gen_sparse` generates sparse tensors for `cpu` and `cuda` devices.

cc @mruberry, @ngimel

aocsa committed:
Step3: Minor refactor and improvements

mberr commented:
@aocsa Did you also run any tests how the performance compares against the [sparse softmax implementation](https://github.com/rusty1s/pytorch_scatter/blob/master/torch_scatter/composite/softmax.py#L7-L25) from [pytorch_scatter](https://github.com/rusty1s/pytorch_scatter)?

aocsa commented:
> @aocsa Did you also run any tests how the performance compares against the [sparse softmax implementation](https://github.com/rusty1s/pytorch_scatter/blob/master/torch_scatter/composite/softmax.py#L7-L25) from [pytorch_scatter](https://github.com/rusty1s/pytorch_scatter)?

@mberr  Let me check pytorch_scatter, indeed it could be interesting to compare both  :-)

aocsa commented:
> > @aocsa Did you also run any tests how the performance compares against the [sparse softmax implementation](https://github.com/rusty1s/pytorch_scatter/blob/master/torch_scatter/composite/softmax.py#L7-L25) from [pytorch_scatter](https://github.com/rusty1s/pytorch_scatter)?
> 
> @mberr Let me check pytorch_scatter, indeed it could be interesting to compare both :-)

Here there are some benchmarks result for `pytorch.sparse.softmax` and  `torch_scatter.softmax` with `dim=0`, [the script is here](https://gist.github.com/aocsa/8a4a4cb9a58d38525db8ba2cc0d16a47), both using CPU and CUDA.

CPU
```
pytorch.sparse.softmax(<(100, 100)-float32-tensor[nnz=1000]>, dim=0) took	 208.80 us
torch_scatter.softmax(<(100, 100)-float32-tensor[nnz=1000]>, dim=0) took	         168.56 us
pytorch.sparse.softmax(<(100, 100)-float32-tensor[nnz=10000]>, dim=0) took	 811.76 us
torch_scatter.softmax(<(100, 100)-float32-tensor[nnz=10000]>, dim=0) took	 306.18 us
pytorch.sparse.softmax(<(1000, 1000)-float32-tensor[nnz=1000]>, dim=0) took	 276.07 us
torch_scatter.softmax(<(1000, 1000)-float32-tensor[nnz=1000]>, dim=0) took	 177.72 us
pytorch.sparse.softmax(<(1000, 1000)-float32-tensor[nnz=10000]>, dim=0) took	 917.53 us
torch_scatter.softmax(<(1000, 1000)-float32-tensor[nnz=10000]>, dim=0) took	 382.92 us
```

CUDA
```
pytorch.sparse.softmax(<(100, 100)-float32-tensor[nnz=1000]>, dim=0) took	 374.56 us
torch_scatter.softmax(<(100, 100)-float32-tensor[nnz=1000]>, dim=0) took	         232.34 us
pytorch.sparse.softmax(<(100, 100)-float32-tensor[nnz=10000]>, dim=0) took	 497.29 us
torch_scatter.softmax(<(100, 100)-float32-tensor[nnz=10000]>, dim=0) took	 390.73 us
pytorch.sparse.softmax(<(1000, 1000)-float32-tensor[nnz=1000]>, dim=0) took	 379.23 us
torch_scatter.softmax(<(1000, 1000)-float32-tensor[nnz=1000]>, dim=0) took	 235.58 us
pytorch.sparse.softmax(<(1000, 1000)-float32-tensor[nnz=10000]>, dim=0) took	 428.70 us
torch_scatter.softmax(<(1000, 1000)-float32-tensor[nnz=10000]>, dim=0) took	 454.08 us
```

Some notes:
- In general, for this case both are comparable, the difference should be because `pytorch.sparse.softmax` creates copies for indices and it covers more general cases likes sparse tensors using a hybrid representation.  
- To extend the use of `torch_scatter.softmax` for sparse tensors with hybrid representation it should require a preprocessing step in order to get the pools. 

Thanks @mberr, it was nice to known about `pytorch_scatter` and how it implements softmax using a scatter operation 
cc @mruberry, @ngimel

ngimel commented:
@aosca your benchmarking script does not properly synchronize cuda (in particular, ~8us for (512,10000) softmax is unrealistic). Here's a script that benchmarks dense implementations, please modify it to include your sparse benchmarks. 
```
import torch
from torch.utils._benchmark import Timer
from torch.utils._benchmark import Compare
import sys


results = []
num_threads = 1
sizes = ((32,10000), (512,10000))
for size in sizes:
    print(size)
    input = torch.randn(size, device="cuda")
    input_cpu = torch.randn(size)

    tasks = [("torch.softmax(input, dim=-1)", "dense CUDA"),
               ("torch.softmax(input_cpu, dim=-1)", "dense CPU")]
    timers = [Timer(stmt=stmt, num_threads=num_threads, label="Softmax", sub_label=f"{size}",
    description=label, globals=globals()) for stmt, label in tasks]
    repeats = 3

    for i, timer in enumerate(timers * repeats):
        results.append(
            timer.blocked_autorange()
        )
        print(f"\r{i + 1} / {len(timers) * repeats}", end="")
        sys.stdout.flush()
print()

comparison = Compare(results)
comparison.print()
```

aocsa commented:
Thanks @ngimel for notice this and the feedback. I updated the [script](https://gist.github.com/aocsa/fbc1827b3e49901512a33ba96092cbc1) and the benchmark results.   I updated the other results as well. 

| size         | density | sparse CUDA | sparse CPU |
|--------------|---------|-------------|------------|
|  (32, 10000) |   0.01  |    380.2    |    687.5   |
| (32, 10000)  | 0.05    | 404.3       | 2357.9     |
| (32, 10000)  | 0.1     | 405.9       | 3677.2     |
| (512, 10000) | 0.01    | 438.0       | 5443.4     |
| (512, 10000) | 0.05    | 888.1       | 24485.0    |
| (512, 10000) | 0.1     | 1921.3      | 45340.5    |


| size         | density | dense CUDA | dense CPU |
|--------------|---------|-------------|------------|
|  (32, 10000) |   0.01  |     23.6    |   1943.2   |
| (32, 10000)  | 0.05    | 23.6        | 1954.0     |
| (32, 10000)  | 0.1     | 23.5        | 1950.0     |
| (512, 10000) | 0.01    | 639.3       | 39797.9    |
| (512, 10000) | 0.05    | 640.3       | 39374.4    |
| (512, 10000) | 0.1     | 639.6       | 39192.3    |

Times are in microseconds (us).

mruberry commented:
Thanks for the performance follow-up, @aocsa. This looks reasonable and consistent with our thinking that sparse is intended for memory savings first and not performance, or at least not until density is sufficiently low and the matrices sufficiently large. @ngimel and I will take another look tomorrow. We'll also be looking at backward. Would you perform the same analysis with backward timings?

aocsa commented:
> Thanks for the performance follow-up, @aocsa. This looks reasonable and consistent with our thinking that sparse is intended for memory savings first and not performance, or at least not until density is sufficiently low and the matrices sufficiently large. @ngimel and I will take another look tomorrow. We'll also be looking at backward. Would you perform the same analysis with backward timings?

Thanks @mruberry, Here it is the [backward timings ](https://gist.github.com/aocsa/e0d15a7bef644d3ae9770ec34af16dbf) for sparse softmax using CPU and CUDA.

| size         | density | sparse CUDA | sparse CPU |
|--------------|---------|-------------|------------|
|(32, 10000) | 0.01   |       88.7    |      125.8 |
|(32, 10000) | 0.05   |       90.4    |      353.4 |
|(32, 10000) | 0.1    |       89.3    |      615.2 |
|(512, 10000) | 0.01  |     1562.3    |    21054.6 |
|(512, 10000) | 0.05  |     1767.1    |    70732.9 |
|(512, 10000) | 0.1   |     2581.1    |   132660.7 |

Let me know if any further changes/updates are necessary. 
Thanks
Alexander

mruberry commented:
> > Thanks for the performance follow-up, @aocsa. This looks reasonable and consistent with our thinking that sparse is intended for memory savings first and not performance, or at least not until density is sufficiently low and the matrices sufficiently large. @ngimel and I will take another look tomorrow. We'll also be looking at backward. Would you perform the same analysis with backward timings?
> 
> Thanks @mruberry, Here it is the [backward timings ](https://gist.github.com/aocsa/e0d15a7bef644d3ae9770ec34af16dbf) for sparse softmax using CPU and CUDA.
> 
> size	density	sparse CUDA	sparse CPU
> (32, 10000)	0.01	88.7	125.8
> (32, 10000)	0.05	90.4	353.4
> (32, 10000)	0.1	89.3	615.2
> (512, 10000)	0.01	1562.3	21054.6
> (512, 10000)	0.05	1767.1	70732.9
> (512, 10000)	0.1	2581.1	132660.7
> Let me know if any further changes/updates are necessary.
> Thanks
> Alexander

How does that compare with dense backward timings?

aocsa commented:
> > > Thanks for the performance follow-up, @aocsa. This looks reasonable and consistent with our thinking that sparse is intended for memory savings first and not performance, or at least not until density is sufficiently low and the matrices sufficiently large. @ngimel and I will take another look tomorrow. We'll also be looking at backward. Would you perform the same analysis with backward timings?
> > 
> > 
> > Thanks @mruberry, Here it is the [backward timings ](https://gist.github.com/aocsa/e0d15a7bef644d3ae9770ec34af16dbf) for sparse softmax using CPU and CUDA.
> > size	density	sparse CUDA	sparse CPU
> > (32, 10000)	0.01	88.7	125.8
> > (32, 10000)	0.05	90.4	353.4
> > (32, 10000)	0.1	89.3	615.2
> > (512, 10000)	0.01	1562.3	21054.6
> > (512, 10000)	0.05	1767.1	70732.9
> > (512, 10000)	0.1	2581.1	132660.7
> > Let me know if any further changes/updates are necessary.
> > Thanks
> > Alexander
> 
> How does that compare with dense backward timings?

Ohh sorry I forgot to paste these timings. Here it is. 

> **softmax backward**

|     size     | density | sparse CUDA | sparse CPU | dense CUDA | dense CPU |
|:------------:|:-------:|:-----------:|:----------:|------------|-----------|
| (32, 10000)  | 0.01    | 88.7        | 125.8      |    13.7    |   136.2   |
| (32, 10000)  | 0.05    | 90.4        | 353.4      | 13.9       | 135.9     |
| (32, 10000)  | 0.1     | 89.3        | 615.2      | 13.8       | 136.3     |
| (512, 10000) | 0.01    | 1562.3      | 21054.6    | 828.6      | 54761.1   |
| (512, 10000) | 0.05    | 1767.1      | 70732.9    | 827.9      | 51361.0   |
| (512, 10000) | 0.1     | 2581.1      | 132660.7   | 829.4      | 51562.3   |

cc @mruberry

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+
+  auto pool_size = pool_offsets.size(0);
+
+  auto exp_sums_rows = at::empty({nvalues * pool_size}, values.options());
```
at::empty -> at::zeros should let you get rid of the thrust::fill?

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+    int64_t* pool_indices = sorted_pool_indices + offset;
+    int64_t pool_indices_size = pool_sizes[index];
+
+    scalar_t* tmp_row = exp_sums_rows + index * nvalues;
```
For exp_sums_rows, it looks like:

- each element of exp_sums_rows is only accessed by a single thread
- the value of exp_sums_rows after the kernel is run is irrelevant

This goes back to an earlier question: why is it needed vs. just creating a local array of size nvalues initialized to zero?

mruberry replied:
(Basically: can we port the improvements you made to forward to backward.)

aocsa replied:
All right

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+    int64_t nvalues,
+    scalar_t* mx_rows,
+    PackedTensorAccessor<scalar_t, 2> input_values_acc,
+    PackedTensorAccessor<scalar_t, 2> ouput_values_acc) {
```
ouput -> output

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+      auto j = lower_bound_values[i];
+      if (j < grad_nnz && (out_offsets[i] == grad_offsets[j])) {
+        auto grad_values_row = grad_values_accessor[j];
+        for (int64_t k = 0; k < nvalues; k++) {
```
Following the improvements in forward, should this be the outer loop?

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+      auto values_row = values_accessor[i];
+      auto j = lower_bound_values[i];
+
+      if (j < grad_nnz && (out_offsets[i] == grad_offsets[j])) {
```
Please add a comment clarifying the purpose of this if statement

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+    for (int64_t p = 0; p < pool_indices_size; p++) {
+      auto i = pool_indices[p];
+      auto out_values_row = out_values_accessor[i];
+      auto values_row = values_accessor[i];
```
values_row doesn't appear to be used in this loop - can you get rid of it here?

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  auto nnz = indices.size(1);
+  auto offsets  = get_offsets(indices, sizes, dim);
```
There are two spaces before your equal sign here

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+        out_offsets.to(at::Device(kCPU), indices.dtype(), false, true);
+    auto host_grad_offsets =
+        grad_offsets.to(at::Device(kCPU), indices.dtype(), false, true);
+    ;
```
stray semicolon

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+  if (dim >= sparse_dim) {
```
Please add a comment here explaining these two code paths

mruberry replied:
(e.g. "when dim >= sparse_dim the dense backward is used")

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+    ;
+    auto out_offsets_accessor = host_out_offsets.data_ptr<int64_t>();
+    auto grad_offsets_accessor = host_grad_offsets.data_ptr<int64_t>();
+    for (int64_t i = 0; i < out_nnz; i++) {
```
Quick question: does the existing sparse softmax test trigger these dense as well as the sparse code paths?

aocsa replied:
Yes, it does. This case is tested when we generate sparse tensors with hybrid representation as well we test with all possible input dim values for `pytorch.sparse.softmax(input, dim)`.

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+          grad_offsets_accessor + grad_offsets.size(0),
+          out_offsets_accessor[i]);
+      auto j = low - grad_offsets_accessor;
+      if (j < grad_nnz && out_offsets_accessor[i] == grad_offsets_accessor[j]) {
```
Same question about the test here: are both these code paths triggered?

aocsa replied:
Here we compute the output using dense backward only when limits and pools are valid, so  this path is also triggered.

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+          grad_offsets_accessor + grad_offsets.size(0),
+          out_offsets_accessor[i]);
+      auto j = low - grad_offsets_accessor;
+      if (j < grad_nnz && out_offsets_accessor[i] == grad_offsets_accessor[j]) {
```
Please add a comment explaining this if check. If this check is false then what happens? Only an empty tensor is returned?

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+          values[i].copy_(r);
+        }
+      }
+    }
```
need an else path here?

mruberry replied:
cc @pearu, too

What happens if a sparse tensor receives a gradient with a different sparsity pattern? Is that an error? Do we check for that consistently?

aocsa replied:
To the best of my knowledge, If this check is false then a sparse tensor with full of zeros (see `values.zero_()` initialization) is returned, so no else path is needed.  cc @pearu

mruberry replied:
Interesting. I guess we don't need to address this issue as long as we're consistent with CPU softmax.

Does the test cover this case?

aocsa replied:
Yes, we're consistent with CPU softmax and it is covered in the tests.

mruberry commented on aten/src/ATen/native/sparse/cuda/SoftMax.cu:
```diff
@@ -0,0 +1,640 @@
+}
+
+} // namespace native
+} // namespace at
```
newline

mruberry commented on test/test_sparse.py:
```diff
@@ -2386,7 +2386,7 @@ def test_sparse_to_numpy(self):
         self.assertRaises(TypeError, lambda: t.numpy())
 
-    @cpu_only
+    @skipIfRocm
```
Why is the skip on ROCm necessary?

aocsa replied:
I skipped ROCM as in other sparse functions  are skipped also.  Just in case let me try .

aocsa replied:
I tested with ROCm enabled and It can't run the second test.  
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.7-py3.6-test2/1903/console

As it is not related specifically to softmax, I will enable again this decorator.

mruberry commented:
@ngimel and I had the opportunity to review again and we think this PR is looking better and better, which is great. We made a few comments about backwards and have some questions about test coverage. Mostly porting the improvements you made in forward to backward and validating that all code paths work as intended.

aocsa committed:
minor updates to softmax backward  functions

aocsa commented:
> @ngimel and I had the opportunity to review again and we think this PR is looking better and better, which is great. We made a few comments about backwards and have some questions about test coverage. Mostly porting the improvements you made in forward to backward and validating that all code paths work as intended.

Thanks for the feedback :+1:, I updated the PR.

mruberry commented:
Your updates look excellent, @aocsa. I know this is inheriting a softmax test from CPU softmax, but how good is that test's coverage? Does it cover the dense code path and its backward, mismatched sparsity patterns, completely sparse and partially dense cases? If not, do you think improving the test coverage is a separable follow-up PR, or are there modifications you'd like to make to the test for this PR?

aocsa committed:
enable again skipIfRocm decorator

ngimel commented:
It would make sense to have a special case where gradient sparsity pattern is the same as output sparsity pattern. As sparse softmax is interesting with masking semantics, this case should be pretty common. Currently, so support possibly different sparsity patterns with mixed sparse-dense tensor, there's a loop calling dense softmax for each nonzero element, it's extremely wasteful. 
Supporting same sparsity pattern for the purely sparse tensors is also important. Currently, if we just randomly generate inputs and gradOutputs (do we?) the gradients even for 10% sparse tensors are mostly zeros, as intersection between out and gradOut is empty. In this case, the backward tests are not even serving their purpose?

pearu commented:
> Currently, if we just randomly generate inputs and gradOutputs (do we?) the gradients even for 10% sparse tensors are mostly zeros, as intersection between out and gradOut is empty. In this case, the backward tests are not even serving their purpose?

@ngimel Here is a list of sparse softmax tests:
https://github.com/pytorch/pytorch/blob/master/test/test_sparse.py#L2651-L2662
As you can see, the test cases are sparse tensors (including hybrid cases) with variable sparsity, few are practically full sparse tensors (when `nnz >=product(shape)`). Moreover, the sparse softmax backward results are tested against full analytical as well as numerical jacobians. So, in my opinion, the tests are pretty thorough here.

pearu commented:
> Currently, so support possibly different sparsity patterns with mixed sparse-dense tensor, there's a loop calling dense softmax for each nonzero element, it's extremely wasteful.

@ngimel, could you provide a link to the code that you are discussing here?

ngimel commented:
> > Currently, so support possibly different sparsity patterns with mixed sparse-dense tensor, there's a loop calling dense softmax for each nonzero element, it's extremely wasteful.
> 
> @ngimel, could you provide a link to the code that you are discussing here?

https://github.com/pytorch/pytorch/pull/42307/files#diff-e4c5af51ba21faf9f0ba3ab58dbf8140R475-R498

pearu commented:
> > > Currently, so support possibly different sparsity patterns with mixed sparse-dense tensor, there's a loop calling dense softmax for each nonzero element, it's extremely wasteful.
> > 
> > 
> > @ngimel, could you provide a link to the code that you are discussing here?
> 
> https://github.com/pytorch/pytorch/pull/42307/files#diff-e4c5af51ba21faf9f0ba3ab58dbf8140R475-R498

Thanks @ngimel ! The referenced (CPU specific) code does not call dense softmax and your note is irrelevant for this PR that deals with CUDA port that has a different implementation.

pearu commented:
> Does it cover the dense code path and its backward, mismatched sparsity patterns, completely sparse and partially dense cases?

@mruberry , yes, except that in the context of softmax, "mismatched sparsity patterns" does not make sense.

ngimel commented:
The code I linked is in SoftMax.cu file, it calls dense cuda softmax backward multiple times (in case of mixed sparse-dense tensor, and softmax dimension being dense). I don't see how it's not relevant for this PR.

pearu commented:
> The code I linked is in SoftMax.cu file, it calls dense cuda softmax backward multiple times (in case of mixed sparse-dense tensor, and softmax dimension being dense). I don't see how it's not relevant for this PR.

@ngimel , the link did not open up SoftMax.cu for me, so I saw only SoftMax.cpp.. You are right, the code in SoftMax.cu for `dim >=  sparse_dim` is suboptimal.

@aocsa , please use the approach in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/SoftMax.cpp#L291-L300 to fix it.

pearu commented:
Hmm, actually, the corresponding CPU code is https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/SoftMax.cpp#L413-L429 which also is a loop over many dense softmax. I am not sure that this can be optimized.

pearu commented:
> I am not sure that this can be optimized, in general.

That said, when `grad_offsets == out_offsets`, the loop over `range(out_nzz)` could be collapsed into a single dense softmax call.

ngimel commented:
> > I am not sure that this can be optimized, in general.
> 
> That said, when `grad_offsets == out_offsets`, the loop over `range(out_nnz)` could be collapsed into a single dense softmax call.

Yes! That's what I meant by "It would make sense to have a special case where gradient sparsity pattern is the same as output sparsity pattern"

aocsa commented:
> > > I am not sure that this can be optimized, in general.
> > 
> > 
> > That said, when `grad_offsets == out_offsets`, the loop over `range(out_nnz)` could be collapsed into a single dense softmax call.
> 
> Yes! That's what I meant by "It would make sense to have a special case where gradient sparsity pattern is the same as output sparsity pattern"

All right, so this optimization is for both Softmax.cpp and Softmax.cu and new tests for this special case. @pearu please could you detail the case here as I am not so familiar with all cases of the backward algorithm, and thanks for your comments, I appreciate your support here. cc @ngimel

aocsa committed:
case: gradient sparsity pattern is the same as output sparsity pattern

minor fix

aocsa commented:
Quick update: I pushed my changes in the PR with the support for the "case: gradient sparsity pattern is the same as output sparsity pattern" when `grad_offsets == out_offsets`. Moreover, I double-checked  if current tests cover this case so there is no necessity to create new ones unless we need to break the tests into more modular tests.  Let me know if you have any other request  and thanks for the feedback @pearu, @ngimel.

mruberry commented:
> Quick update: I pushed my changes in the PR with the support for the "case: gradient sparsity pattern is the same as output sparsity pattern" when `grad_offsets == out_offsets`. Moreover, I double-checked if current tests cover this case so there is no necessity to create new ones unless we need to break the tests into more modular tests. Let me know if you have any other request and thanks for the feedback @pearu, @ngimel.

Thanks @aocsa! We'll take a look soon.

mruberry commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -291,10 +292,10 @@ void cpu_sparse_coo_softmax(Tensor output, const Tensor& input, const int64_t di
     if (LogSoftMax) {
       auto new_values = log_softmax_cpu(values, dim - sparse_dim + 1, false);
-      out_values.copy_(new_values);
+      out_values.set_(new_values);
```
NOTE: this is only OK if there's no user-facing API where the user can provide an out tensor. In those cases we want to preserve the storage.

mruberry reviewed (approved):
Thanks for all the updates, @aocsa. @ngimel and I had the chance to review and we're happy with how this looks.

In the future it'll be interesting to improve our test infrastructure so we don't need to write so many custom components (like the Python softmax implementation or the custom jacobian implementation). That should make adding new operators much easier.

mruberry commented:
Unfortunately this is triggering our arcane internal build system, which is complaining about an inability to find <ATen/native/sparse/utils/ParamUtils.h>. The issue is, I think, that it's in a new folder that I'd have to register internally in a few places. A more immediate fix, however, would be to refactor it into the current sparse folder instead of putting it under utils. Would that be OK for now, @aocsa? We can revisit create a sparse utils folder later if we have organizational issues.

aocsa commented:
> Unfortunately this is triggering our arcane internal build system, which is complaining about an inability to find <ATen/native/sparse/utils/ParamUtils.h>. The issue is, I think, that it's in a new folder that I'd have to register internally in a few places. A more immediate fix, however, would be to refactor it into the current sparse folder instead of putting it under utils. Would that be OK for now, @aocsa? We can revisit create a sparse utils folder later if we have organizational issues.

Sure, I updated the PR with this minor refactor.  cc @mruberry

mruberry commented:
Thank you!

mruberry commented:
Ah, sorry, forgot to mention you'll also have to move the function definitions (implementations) to a cpp file. Only the declarations go in the header.

aocsa committed:
minor refactor ParamUtils


minor refactor

aocsa committed:
minor refactor

aocsa commented:
> Ah, sorry, forgot to mention you'll also have to move the function definitions (implementations) to a cpp file. Only the declarations go in the header.

Sure. It's done.
